dd-sdk-android/src/main/cpp/utils/fileutils.cpp

```diff
@@ -0,0 +1,13 @@
+#include <dirent.h>
+#include <cerrno>
+#include <sys/stat.h>
+#include "fileutils.h"
+
+
+void createDirIfNotExists(const char *dirPath) {
+    opendir(dirPath);
+    if (ENOENT == errno) {
+        // directory does not exist. We will create it.
+        mkdir(dirPath, S_IRWXU);
```

Comment:
Quick question, does `mkdir` create intermediate folders too ?

Reply:
Also not an expert but shouldn't you deal with the return value of `openDIr()`?

Reply:
Nope...I did not try but don't think is doing this recursively but in any case we do not need it here as the 'data' folder will always be there.

Reply:
For the second question it doesn't work like that ...basically the `errno` is an int * to the last error thrown in the stack...and it verifies this with ENOENT which basically means that it could not find that path.

Reply:
Actually you need to check for the opendir() return value as it only sets errno if there is an issue.
Imagine you do: `opendir(path1); [...]; opendir(path2);`: you would have no way to know whether `errno` was set for the first or the second call. So the correct way to write this would be:
```
    if ((opendir(dirPath) == NULL) && (ENOENT == errno)) {
        [...]
    }
```

Reply:
For the recursive part: no, `mkdir` will not create subdirectories recursively. And I must say I am not exactly convinced by your argument of you saying that "it will always exist". If this the case, then check it (i.e. make sure in your code that the parent directory exists) and change the name of the function (something like `createDataSubdirectory()`) so that no one could ever think that this is a generic function like its name implies.

Reply:
And if you want a recursive `mkdir()` implementation here is one ;-)
```
static void _mkdir(const char *dir) {
        char tmp[256];
        char *p = NULL;
        size_t len;
 
        snprintf(tmp, sizeof(tmp),"%s",dir);
        len = strlen(tmp);
        if(tmp[len - 1] == '/') {
                tmp[len - 1] = 0;
        }
        for(p = tmp + 1; *p; p++) {
                if(*p == '/') {
                        *p = 0;
                        mkdir(tmp, S_IRWXU);
                        *p = '/';
                }
        }
        mkdir(tmp, S_IRWXU);
}
```

Reply:
The reason I said the parent dir will always exist is that normally it has to be created by the OS whenever the app is installed with the current rights for this user (application). I guess if I add a check if the STORAGE_DIR exists or it was successfully created that will be enough because it will also cover the case when the parent one does not exist (in this case it will not be able to create the STORAGE_DIR). In any case I don't want to create this directory recursively as I don't want to override something that the OS should do automatically.